samples/guestbook/gcp/deploy.go

```diff
@@ -0,0 +1,159 @@
+// Copyright 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// The deploy program builds the Guestbook server locally and deploys it to
+// GKE.
+
+package main
+
+import (
+	"encoding/json"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"time"
+)
+
+func main() {
+	guestbookDir := flag.String("guestbook_dir", "..", "directory containing the guestbook example")
+	flag.Parse()
+	if err := deploy(*guestbookDir); err != nil {
+		fmt.Fprintln(os.Stderr, "deploy:", err)
+		os.Exit(1)
+	}
+}
+
+func deploy(guestbookDir string) error {
+	tfStatePath := filepath.Join(guestbookDir, "gcp", "terraform.tfstate")
+	keys := []string{"project", "cluster_name", "cluster_zone", "bucket", "database_instance", "database_region", "motd_var_config", "motd_var_name"}
+	tfState := make(map[string]string)
+	for _, k := range keys {
+		cmd := exec.Command("terraform", "output", "-state", tfStatePath, k)
+		outb, err := cmd.CombinedOutput()
+		if err != nil {
+			return fmt.Errorf("running %v: %v: %s", cmd.Args, err, outb)
+		}
+		tfState[k] = string(outb)
+	}
+
+	gcp := gcloud{project: tfState["project"]}
+	tempDir, err := ioutil.TempDir("", "guestbook-k8s-")
+	if err != nil {
+		return fmt.Errorf("making temp dir: %v", err)
+	}
+	defer os.RemoveAll(tempDir)
+
+	// Fill in Kubernetes template parameters.
+	proj := strings.Replace(tfState["project"], ":", "/", -1)
+	imageName := fmt.Sprintf("gcr.io/%s/guestbook", proj)
+	gbyin, err := ioutil.ReadFile(filepath.Join(guestbookDir, "gcp", "guestbook.yaml.in"))
+	if err != nil {
+		return fmt.Errorf("reading guestbook.yaml.in: %v", err)
+	}
+	gby := string(gbyin)
+	replacements := map[string]string{
```

Comment:
This can be done more simply and robustly with [`strings.NewReplacer`](https://godoc.org/strings#NewReplacer).